File: algorithms/python/206_reverse_linked_list/reverse_linked_list.py
"""
206. reverse linked list

https://leetcode.com/problems/reverse-linked-list/

反转链表
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def reverse_list(head: ListNode) -> ListNode:
    """
    iterative
    """
    prev = None
    while head:
        tmp = head.next
        head.next = prev
        prev = head
        head = tmp

    return prev

File: algorithms/206-reverse-linked-list/reverse-linked-list.go
package problem206

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		tmp := head.Next
		head.Next = prev
		prev = head
		head = tmp
	}

	return prev
}

// 递归
func reverseListV2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	nextHead := reverseListV2(head.Next)
	head.Next.Next = head
	head.Next = nil
	return nextHead
}
